File: Behavior/State/Go/state.go
package main

import "fmt"

// State 接口
type State interface {
	Handle(c *Context)
}

// Context 结构体，代表拥有状态的实体
type Context struct {
	State State
}

// 实现状态转换的方法
func (c *Context) Request() {
	c.State.Handle(c)
}

// 具体状态: StartState
type StartState struct{}

func (s *StartState) Handle(c *Context) {
	fmt.Println("Player is in start state")
	c.State = &StopState{}
}

// 具体状态: StopState
type StopState struct{}

func (s *StopState) Handle(c *Context) {
	fmt.Println("Player is in stop state")
	// 这里不再转换状态，表示结束
}

func main() {
	context := &Context{State: &StartState{}}

	// 应用初始状态的行为
	context.Request()

	// 状态从 StartState 转变到 StopState
	context.Request()

	// 再次调用，但状态是 StopState，应该没有变化
	context.Request()
}

File: Behavior/State/Cpp/state.cpp
#include <iostream>
#include <memory>

// 前向声明
class Context;

// State 接口
class State
{
public:
    virtual void handle(Context &context) = 0;
    virtual ~State() = default;
};

// Context 类
class Context
{
private:
    std::unique_ptr<State> state_;

public:
    void setState(std::unique_ptr<State> state)
    {
        state_ = std::move(state);
    }
    void request()
    {
        state_->handle(*this);
    }
};

// 具体状态类：Finished
class Finished : public State
{
public:
    void handle(Context &context) override
    {
        std::cout << "Finished state. No further state transitions." << std::endl;
    }
};

// 具体状态类：InProgress
class InProgress : public State
{
public:
    void handle(Context &context) override
    {
        std::cout << "InProgress state. Changing state to Finished." << std::endl;
        context.setState(std::make_unique<Finished>());
    }
};

// 具体状态类：Started
class Started : public State
{
public:
    void handle(Context &context) override
    {
        std::cout << "Started state. Changing state to InProgress." << std::endl;
        context.setState(std::make_unique<InProgress>());
    }
};

int main()
{
    Context context;

    // 设置初始状态为 Started
    context.setState(std::make_unique<Started>());

    // 处理请求，状态从 Started -> InProgress
    context.request();

    // 处理请求，状态从 InProgress -> Finished
    context.request();

    // 处理请求，状态为 Finished，不再变化
    context.request();

    return 0;
}
